packages/plugin-react/src/index.ts

```diff
@@ -378,59 +312,7 @@ export default function viteReact(opts: Options = {}): PluginOption[] {
     },
   }
 
-  const reactJsxRuntimeId = 'react/jsx-runtime'
-  const reactJsxDevRuntimeId = 'react/jsx-dev-runtime'
-  const virtualReactJsxRuntimeId = '\0' + reactJsxRuntimeId
-  const virtualReactJsxDevRuntimeId = '\0' + reactJsxDevRuntimeId
-  // Adapted from https://github.com/alloc/vite-react-jsx
-  const viteReactJsx: Plugin = {
-    name: 'vite:react-jsx',
-    enforce: 'pre',
-    config() {
-      return {
-        optimizeDeps: {
-          // We can't add `react-dom` because the dependency is `react-dom/client`
-          // for React 18 while it's `react-dom` for React 17. We'd need to detect
-          // what React version the user has installed.
-          include: [reactJsxRuntimeId, reactJsxDevRuntimeId, 'react'],
-        },
-      }
-    },
-    resolveId(id, importer) {
-      // Resolve runtime to a virtual path to be interoped.
-      // Since the interop code re-imports `id`, we need to prevent re-resolving
-      // to the virtual id if the importer is already the virtual id.
-      if (id === reactJsxRuntimeId && importer !== virtualReactJsxRuntimeId) {
-        return virtualReactJsxRuntimeId
-      }
-      if (
-        id === reactJsxDevRuntimeId &&
-        importer !== virtualReactJsxDevRuntimeId
-      ) {
-        return virtualReactJsxDevRuntimeId
-      }
-    },
-    load(id) {
-      // Apply manual interop
-      if (id === virtualReactJsxRuntimeId) {
-        return [
-          `import * as jsxRuntime from ${JSON.stringify(reactJsxRuntimeId)}`,
-          `export const Fragment = jsxRuntime.Fragment`,
-          `export const jsx = jsxRuntime.jsx`,
-          `export const jsxs = jsxRuntime.jsxs`,
-        ].join('\n')
-      }
-      if (id === virtualReactJsxDevRuntimeId) {
-        return [
-          `import * as jsxRuntime from ${JSON.stringify(reactJsxDevRuntimeId)}`,
-          `export const Fragment = jsxRuntime.Fragment`,
-          `export const jsxDEV = jsxRuntime.jsxDEV`,
-        ].join('\n')
-      }
```

Comment:
Oh ok I get the point, I will take a closer look at this

Reply:
I just looked at the output, feels ok for the jsx naming. The import line is a bit long but not sure anybody look at this.
And this would be fixed when React _finally_ ships esm.
```js
import { injectQuery as __vite__injectQuery } from "/@vite/client";import { createHotContext as __vite__createHotContext } from "/@vite/client";import.meta.hot = __vite__createHotContext("/App.jsx");import __vite__cjsImport0_react_jsxDevRuntime from "/node_modules/.vite/deps/react_jsx-dev-runtime.js?v=88111a26"; const jsxDEV = __vite__cjsImport0_react_jsxDevRuntime["jsxDEV"];
import RefreshRuntime from "/@react-refresh";
let prevRefreshReg;
let prevRefreshSig;
if (import.meta.hot) {
  if (!window.__vite_plugin_react_preamble_installed__) {
    throw new Error("@vitejs/plugin-react can't detect preamble. Something is wrong. See https://github.com/vitejs/vite-plugin-react/pull/11#discussion_r430879201");
  }
  prevRefreshReg = window.$RefreshReg$;
  prevRefreshSig = window.$RefreshSig$;
  window.$RefreshReg$ = (type, id) => {
    RefreshRuntime.register(type, "/Users/arnaud/git/vite-plugin-react/playground/react/App.jsx " + id);
  };
  window.$RefreshSig$ = RefreshRuntime.createSignatureFunctionForTransform;
}
var _s = $RefreshSig$();
import __vite__cjsImport3_react from "/node_modules/.vite/deps/react.js?v=b9578cd9"; const useState = __vite__cjsImport3_react["useState"];
import Button from "/@fs/Users/arnaud/git/vite-plugin-react/node_modules/.pnpm/file+playground+react+jsx-entry/node_modules/jsx-entry/Button.jsx";
import Dummy from "/components/Dummy.jsx?qs-should-not-break-plugin-react";
import Parent from "/hmr/parent.jsx";
import { CountProvider } from "/context/CountProvider.jsx";
import { ContextButton } from "/context/ContextButton.jsx";
function App() {
  _s();
  const [count, setCount] = useState(0);
  return /* @__PURE__ */ jsxDEV("div", { className: "App", children: [
    /* @__PURE__ */ jsxDEV("header", { className: "App-header", children: [
      /* @__PURE__ */ jsxDEV("h1", { children: "Hello Vite + React" }, void 0, false, {
        fileName: "/Users/arnaud/git/vite-plugin-react/playground/react/App.jsx",
        lineNumber: 13,
        columnNumber: 9
      }, this),
      /* @__PURE__ */ jsxDEV("p", { children: /* @__PURE__ */ jsxDEV("button", { id: "state-button", onClick: () => setCount((count2) => count2 + 1), children: [
        "count is: ",
        count
      ] }, void 0, true, {
        fileName: "/Users/arnaud/git/vite-plugin-react/playground/react/App.jsx",
        lineNumber: 15,
        columnNumber: 11
      }, this) }, void 0, false, {
        fileName: "/Users/arnaud/git/vite-plugin-react/playground/react/App.jsx",
        lineNumber: 14,
        columnNumber: 9
      }, this),
      /* @__PURE__ */ jsxDEV("p", { children: /* @__PURE__ */ jsxDEV(ContextButton, {}, void 0, false, {
        fileName: "/Users/arnaud/git/vite-plugin-react/playground/react/App.jsx",
        lineNumber: 20,
        columnNumber: 11
      }, this) }, void 0, false, {
        fileName: "/Users/arnaud/git/vite-plugin-react/playground/react/App.jsx",
        lineNumber: 19,
        columnNumber: 9
      }, this),
      /* @__PURE__ */ jsxDEV("p", { children: [
        "Edit ",
        /* @__PURE__ */ jsxDEV("code", { children: "App.jsx" }, void 0, false, {
          fileName: "/Users/arnaud/git/vite-plugin-react/playground/react/App.jsx",
          lineNumber: 23,
          columnNumber: 16
        }, this),
        " and save to test HMR updates."
      ] }, void 0, true, {
        fileName: "/Users/arnaud/git/vite-plugin-react/playground/react/App.jsx",
        lineNumber: 22,
        columnNumber: 9
      }, this),
      /* @__PURE__ */ jsxDEV("a", { className: "App-link", href: "https://reactjs.org", target: "_blank", rel: "noopener noreferrer", children: "Learn React" }, void 0, false, {
        fileName: "/Users/arnaud/git/vite-plugin-react/playground/react/App.jsx",
        lineNumber: 25,
        columnNumber: 9
      }, this)
    ] }, void 0, true, {
      fileName: "/Users/arnaud/git/vite-plugin-react/playground/react/App.jsx",
      lineNumber: 12,
      columnNumber: 7
    }, this),
    /* @__PURE__ */ jsxDEV(Dummy, {}, void 0, false, {
      fileName: "/Users/arnaud/git/vite-plugin-react/playground/react/App.jsx",
      lineNumber: 30,
      columnNumber: 7
    }, this),
    /* @__PURE__ */ jsxDEV(Parent, {}, void 0, false, {
      fileName: "/Users/arnaud/git/vite-plugin-react/playground/react/App.jsx",
      lineNumber: 31,
      columnNumber: 7
    }, this),
    /* @__PURE__ */ jsxDEV(Button, { children: "button" }, void 0, false, {
      fileName: "/Users/arnaud/git/vite-plugin-react/playground/react/App.jsx",
      lineNumber: 32,
      columnNumber: 7
    }, this)
  ] }, void 0, true, {
    fileName: "/Users/arnaud/git/vite-plugin-react/playground/react/App.jsx",
    lineNumber: 11,
    columnNumber: 10
  }, this);
}
_s(App, "oDgYfYHkD9Wkv4hrAPCkI/ev3YU=");
_c = App;
```

Reply:
Interesting, yes, this looks fine to me too 🤔
Maybe there was another issue trying to be solved, but if the plugin works without this extra complexity it is worth trying it in the major beta, and you could revert later. Maybe it isn't bad to break this PR into a few parts (removing this being one for example) so it is easier to reference later.

Reply:
It was originally added for more minifiable `vite build` output (using `jsx(...)` instead of something like `jsxRuntime.jsx(...)`).

Reply:
Just tested and the build output is ok IMO:
`F.jsxs("p",{children:["Edit ",F.jsx("code",{children:"App.jsx"})," and save to test HMR updates."]})`

Reply:
`F.jsxs` vs `F` is an 83% difference, which is pretty significant if you have a large app. Even though network compression removes the issue for delivery speed, there's still parsing performance. That said, I haven't benchmarked the difference in a large app, so it's only conjecture from my end.

Reply:
yeah gzip size difference should be really small. The extra parsing time/execution time seems like a bit over optimisation for me. I just hope the React team ships esm before 2030